tools/dataset.cpp

```diff
@@ -1 +1,1305 @@
+#include <fcntl.h>     // `open` files
+#include <sys/stat.h>  // `stat` to obtain file metadata
+#include <sys/mman.h>  // `mmap` to read datasets faster
+#include <uuid/uuid.h> // `uuid` to make file name
 
+#include <vector>
+#include <fstream>
+#include <cstring>
+#include <numeric>
+#include <algorithm>
+#include <filesystem>
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wall"
+#pragma GCC diagnostic ignored "-Wextra"
+#include <arrow/api.h>
+#include <arrow/array.h>
+#include <arrow/table.h>
+#include <arrow/result.h>
+#include <arrow/status.h>
+#include <arrow/io/api.h>
+#include <arrow/csv/api.h>
+#include <arrow/io/file.h>
+#include <arrow/csv/writer.h>
+#include <arrow/memory_pool.h>
+#include <parquet/arrow/reader.h>
+#include <parquet/stream_writer.h>
+#include <arrow/compute/api_aggregate.h>
+#pragma GCC diagnostic pop
+
+#include <simdjson.h>
+#include <fmt/format.h>
+
+#include <ukv/ukv.hpp>
+
+#include "dataset.h"
+#include "../benchmarks/mixed.hpp"
+
+#include <ukv/cpp/ranges.hpp>      // `sort_and_deduplicate`
+#include <ukv/cpp/blobs_range.hpp> // `keys_stream_t`
+
+using namespace unum::ukv::bench;
+using namespace unum::ukv;
+
+constexpr std::size_t uuid_length = 36;
+
+using graph_t = std::vector<edge_t>;
+using docs_t = std::vector<value_view_t>;
+using fields_t = strided_iterator_gt<ukv_str_view_t const>;
+
+using ids_t = std::vector<std::pair<ukv_key_t*, ukv_size_t>>;
+using vals_t = std::vector<std::pair<ukv_bytes_ptr_t, ukv_size_t>>;
+
+/////////  Helpers  /////////
+
+class arrow_visitor {
+  public:
+    arrow_visitor(std::string& json) : json(json) {}
+    arrow::Status Visit(arrow::NullArray const& arr) {
+        return arrow::Status(arrow::StatusCode::TypeError, "Not supported type");
+    }
+    arrow::Status Visit(arrow::BooleanArray const& arr) {
+        json = fmt::format("{}{},", json, arr.Value(idx));
```

Comment:
This is horrible. 
You are reallocating a new string every time you want to append a boolean.
At least use `fmt::format_to`.